pipeline/subtractor.py

```diff
@@ -0,0 +1,122 @@
+
+import sqlalchemy as sa
+
+from pipeline.parameters import Parameters
+
+from pipeline.utils import get_image_cache
+
+from models.image import Image, CoaddImage, SubtractionImage
+from models.base import SmartSession
+
+
+class ParsSubtractor(Parameters):
+    def __init__(self, **kwargs):
+        super().__init__()
+        self.algorithm = self.add_par(
+            'algorithm',
+            'hotpants',
+            str,
+            'Which algorithm to use. Possible values are: "hotpants", "zogy". '
+        )
+
+        self._enforce_no_new_attrs = True
+
+        self.override(kwargs)
+
+    def _get_process_name(self):
+        return 'subtraction'
+
+    def _get_upstream_process_names(self):
+        return ['preprocess', 'astrometry', 'calibration']
+
+
+class Subtractor:
+    def __init__(self, **kwargs):
+        self.pars = ParsSubtractor()
+        self.pars.update(kwargs)
+
+        # TODO: add a reference cache here.
```

Comment:
I thought about this, but I would like to separate the constructor from the data flow. 
The constructor is where you set up the `Parameters` object. Then when the pipeline is ready, you give it data. 
This lets you loop on an object without having to re-construct it every time (which could be expensive if the object also caches stuff like reference images). 

I think the choice of reference image should probably be done using provenance choices, but I can also see someone just making a reference locally and passing it in through the `image_cache` (or `context`) to use that explicitly. 
In either case you must have a provenance to the reference so it gets saved in the subtracted image's provenance's upstream.

Reply:
Why would you have to reconstruct it every time if you passed the context in with the constructor?  Would you want to wipe out the image cache between subsequent calls?

For that matter -- why are we calling it more than once in a go through the pipeline?

Reply:
I was thinking something like:
```
pipeline = Pipeline(arguments, ...)

for i in range(...):
  for j in range(...):
    pipeline.run(i, j)
```

So that the same object can be re-used in a loop. If you want, you can initialize it once per image, but you don't have to.

Reply:
I would make it the standard to reinitialize once per image.  Reason: you're talking about caching all of the images in memory.  If you keep that cache around, it's going to keep growing every time you start a new image.  To avoid that becoming a problem, you're going to have to write cache management code, and that's not worth the effort.  It's much easier to just say when you initialize a new subtractor, you specify exactly what image it's for.

The overhead in creating these things, and reading these files from disk, is going to be utterly tiny compared to the time it takes to actually process them.  (The overhead of getting them over the network may well not be, but that's why we have local storage.)

Right now, the way I run the code, I start a whole new executable every time I'm starting a new image.  Again, the run time is minutes, so a few seconds of startup time is not a significant hit.

It will simplify how we deal with things if we specify the context / image etc. in the constructor, and not have to worry about cache bloat and such.